x/valset-pref/validator_set.go

```diff
@@ -29,6 +29,44 @@ func (k Keeper) SetupValidatorSetPreference(ctx sdk.Context, delegator string, p
 	return nil
 }
 
+// TODO MAYBE: Check if there are any banned assets and ways to handle them
+// DelegateToValidatorSet delegates to a delegators existing validator-set.
+// For ex: delegate 10osmo with validator-set {ValA -> 0.5, ValB -> 0.3, ValC -> 0.2}
+// our delegate logic would attempt to delegate 5osmo to A , 2osmo to B, 3osmo to C
+func (k Keeper) DelegateToValidatorSet(ctx sdk.Context, delegatorAddr string, coin sdk.Coin) error {
+	// get the existing validator set preference from store
+	existingSet, found := k.GetValidatorSetPreference(ctx, delegatorAddr)
+	if !found {
+		return fmt.Errorf("user %s doesn't have validator set", delegatorAddr)
+	}
+
+	delegator, err := sdk.AccAddressFromBech32(delegatorAddr)
+	if err != nil {
+		return err
+	}
+
+	// loop through the validatorSetPreference and delegate the proportion of the tokens based on weights
+	for _, val := range existingSet.Preferences {
+		_, validator, err := k.getValAddrAndVal(ctx, val.ValOperAddress)
+		if err != nil {
+			return err
+		}
+
+		// tokenAmt takes the amount to delegate, calculated by {val_distribution_weight * tokenAmt}
+		// expect tokenAmt to be in the smallest denom for ex, 1 usmo = 10^6 osmo
+		tokenAmt := val.Weight.Mul(coin.Amount.ToDec()).RoundInt()
+
+		// TODO: What happens here if validator is jailed, tombstoned, or unbonding
+		// Delegate the unbonded tokens
+		_, err = k.stakingKeeper.Delegate(ctx, delegator, tokenAmt, stakingtypes.Unbonded, validator, true)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
The only issue I see with this is if delegating to a tombstoned validator produces an error. If this is the case, If we are halfway through the loop and we hit a tombstoned validator, it seems like we would error and the remainder of the delegation doesn't get delegated. There should be some kind of contract here that we either delegate all or none. Should check if delegating to tombstoned validators produces error. If it does, we need to check if its tombstoned prior to this. If not, can probably leave as is.

Reply:
hmm one option would be to loop through the val-set before delegation and check if there are any jailed / invalid validators or not using `[IsValidatorJailed](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/validator.go#L480)`  or [GetValidator](https://github.com/cosmos/cosmos-sdk/blob/main/x/staking/keeper/validator.go#L15) 

And if there is any jailed validator we DONOT delegate anything. lmk what you think

Reply:
@mattverse we talked about this in the chain dev call. If we loop through the validators and check their jail, tombstoned &  unbonding status it might be very gas intensive, therefore i believe we can just have simple validator check like `GetValidator` and leave this as is. 

Another reason why we are being lenient with this check is that if in any iteration of the loop the validator is faulty, then the function returns an error and the entire message is reverted. **We are doing many iterations of `Delegate` function using only 1 message so we might have a leeway here**

Reply:
Just to confirm, that means that we're still adding a check to check if a validator is jailed, and if so, reverts the entire message?

Reply:
we have `GetValidator` that runs in ever iteration of `Val-Set` which checks that